services/motion/explore/explore.go

```diff
@@ -0,0 +1,575 @@
+	ctx context.Context,
+	req motion.ListPlanStatusesReq,
+) ([]motion.PlanStatusWithID, error) {
+	return nil, errUnimplemented
+}
+
+func (ms *explore) PlanHistory(
+	ctx context.Context,
+	req motion.PlanHistoryReq,
+) ([]motion.PlanWithStatus, error) {
+	return nil, errUnimplemented
+}
+
+func (ms *explore) Close(ctx context.Context) error {
+	utils.FlushChan(ms.obstacleResponseChan)
+	utils.FlushChan(ms.executionResponseChan)
+	ms.backgroundWorkers.Wait()
+	return nil
+}
+
+// Move takes a goal location and will plan and execute a movement to move a component specified by its name
+// to that destination.
+func (ms *explore) Move(
+	ctx context.Context,
+	componentName resource.Name,
+	destination *referenceframe.PoseInFrame,
+	worldState *referenceframe.WorldState,
+	constraints *servicepb.Constraints,
+	extra map[string]interface{},
+) (bool, error) {
+	operation.CancelOtherWithLabel(ctx, exploreOpLabel)
+
+	// Parse extras
+	motionCfg, err := parseMotionConfig(extra)
+	if err != nil {
+		return false, err
+	}
+
+	// obstacleDetectors
+	obstacleDetectors, err := ms.createObstacleDetectors(motionCfg)
+	if err != nil {
+		return false, err
+	}
+
+	// Create kinematic base
+	kb, err := ms.createKinematicBase(ctx, componentName, motionCfg)
+	if err != nil {
+		return false, err
+	}
+
+	// Create motionplan plan
+	planInputs, err := ms.createMotionPlan(ctx, kb, destination.Pose(), worldState, extra)
+	if err != nil {
+		return false, err
+	}
+	var plan motionplan.Plan
+	for _, inputs := range planInputs {
+		input := make(map[string][]referenceframe.Input)
+		input[kb.Name().Name] = inputs
+		plan = append(plan, input)
+	}
+
+	// Start background processes
+	cancelCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Start polling for obstacles
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.checkForObstacles(cancelCtx, obstacleDetectors, kb, plan, motionCfg.ObstaclePollingFreqHz)
+	}, ms.backgroundWorkers.Done)
+
+	// Start executing plan
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.executePlan(cancelCtx, kb, plan)
+	}, ms.backgroundWorkers.Done)
+
+	for {
+		// this ensures that if the context is cancelled we always return early at the top of the loop
+		if err := ctx.Err(); err != nil {
+			return false, err
+		}
+
+		select {
+		// if context was cancelled by the calling function, error out
+		case <-ctx.Done():
+			return false, ctx.Err()
+
+		// once execution responds: return the result to the caller
+		case resp := <-ms.executionResponseChan:
+			ms.logger.Debugf("execution completed: %s", resp)
+			return resp.success, resp.err
+
+		// if the checkPartialPlan process hit an error return it, otherwise exit
+		case resp := <-ms.obstacleResponseChan:
+			ms.logger.Debugf("obstacle response: %s", resp)
+			if resp.err != nil {
+				return resp.success, resp.err
+			}
+			if resp.success {
+				return resp.success, nil
+			}
+		}
+	}
+}
+
+type moveResponse struct {
```

Comment:
[opt]: prefer to have structs defined at the top of the file

Reply:
Moved up and comment added